cmd/whitesourceExecuteScan.go

```diff
@@ -291,9 +574,125 @@ func writeWhitesourceConfigJSON(config *ScanOptions, utils whitesourceUtils, dev
+		if err != nil {
+			return fmt.Errorf("failed to scan NPM module '%s': %w", module, err)
+		}
+	}
+	return nil
+}
+
+// executeNpmScanForModule generates a configuration file whitesource.config.json with appropriate values from config,
+// installs all dependencies if necessary, and executes the scan via "npx whitesource run".
+func executeNpmScanForModule(modulePath string, config *ScanOptions, scan *whitesourceScan, utils whitesourceUtils) error {
+	log.Entry().Infof("Executing Whitesource scan for NPM module '%s'", modulePath)
+
+	resetDir, err := utils.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to obtain current directory: %w", err)
+	}
+
+	dir := filepath.Dir(modulePath)
+	if err := utils.Chdir(dir); err != nil {
+		return fmt.Errorf("failed to change into directory '%s': %w", dir, err)
+	}
+	defer func() {
+		err = utils.Chdir(resetDir)
+		if err != nil {
+			log.Entry().Errorf("Failed to reset into directory '%s': %v", resetDir, err)
+		}
+	}()
+
+	if err := writeWhitesourceConfigJSON(config, utils, false, true); err != nil {
+		return err
+	}
+	defer func() { _ = utils.FileRemove(whiteSourceConfig) }()
+
+	projectName, err := getNpmProjectName(modulePath, utils)
+	if err != nil {
+		return err
+	}
+
+	if err := reinstallNodeModulesIfLsFails(modulePath, config, utils); err != nil {
+		return err
+	}
+
+	if err := scan.appendScannedProject(projectName); err != nil {
+		return err
+	}
+
+	return utils.RunExecutable("npx", "whitesource", "run")
+}
+
+func getNpmProjectName(modulePath string, utils whitesourceUtils) (string, error) {
+	fileContents, err := utils.FileRead("package.json")
+	if err != nil {
+		return "", fmt.Errorf("could not read package.json: %w", err)
+	}
+	var packageJSON = make(map[string]interface{})
+	err = json.Unmarshal(fileContents, &packageJSON)
+
+	projectNameEntry, exists := packageJSON["name"]
+	if !exists {
+		return "", fmt.Errorf("the file '%s' must configure a name",
+			filepath.Join(modulePath, "package.json"))
+	}
+
+	projectName, isString := projectNameEntry.(string)
+	if !isString {
+		return "", fmt.Errorf("the file '%s' must configure a name",
+			filepath.Join(modulePath, "package.json"))
+	}
+
+	return projectName, nil
+}
+
+func reinstallNodeModulesIfLsFails(modulePath string, config *ScanOptions, utils whitesourceUtils) error {
+	// No need to have output from "npm ls" in the log
+	utils.Stdout(ioutil.Discard)
+	defer utils.Stdout(log.Writer())
+
+	err := utils.RunExecutable("npm", "ls")
+	if err == nil {
+		return nil
+	}
+	log.Entry().Warnf("'npm ls' failed. Re-installing NPM Node Modules")
+	err = utils.RemoveAll("node_modules")
+	if err != nil {
+		return fmt.Errorf("failed to remove node_modules directory: %w", err)
+	}
+	err = utils.MkdirAll("node_modules", os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("failed to recreate node_modules directory: %w", err)
+	}
+	exists, _ := utils.FileExists("package-lock.json")
+	if exists {
+		err = utils.FileRemove("package-lock.json")
```

Comment:
I know we are currently doing it, but it is kind of a dirty hack which we should not do anymore ...

Reply:
I saw that the previous implementation for the npm scan passed a parameter `"ignoreNpmLsErrors":true` in the generated configuration. Our old implementation passed `"ignoreNpmLsErrors":false`, but implemented this check for "npm ls". Should we simply remove this code and always pass `"ignoreNpmLsErrors":true`? In our test-project, this check triggers for one of the npm-modules, so I know the code works. But I can test what happens if I flip the option.

Reply:
I would vote for "ignoreNpmLsErrors":true and remove this hack....